subsys/net/lib/lwm2m/lwm2m_message_handling.c

```diff
@@ -1045,19 +1051,27 @@ int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst, struct lwm2m_eng
 					     res_inst->res_inst_id, &data_len);
 	}
 
-	if (res->post_write_cb
-#if CONFIG_LWM2M_ENGINE_VALIDATION_BUFFER_SIZE > 0
-	    || res->validate_cb
-#endif
-	) {
-		if (msg->in.block_ctx != NULL) {
-			/* Get block_ctx for total_size (might be zero) */
-			total_size = msg->in.block_ctx->ctx.total_size;
-			LOG_DBG("BLOCK1: total:%zu current:%zu"
-				" last:%u",
-				msg->in.block_ctx->ctx.total_size, msg->in.block_ctx->ctx.current,
-				msg->in.block_ctx->last_block);
+	if (msg->in.block_ctx != NULL) {
+		/* Get block_ctx for total_size (might be zero) */
+		total_size = msg->in.block_ctx->ctx.total_size;
+		offset = msg->in.block_ctx->ctx.current;
+
+		LOG_DBG("BLOCK1: total:%zu current:%zu"
+			" last:%u",
+			msg->in.block_ctx->ctx.total_size, msg->in.block_ctx->ctx.current,
+			msg->in.block_ctx->last_block);
+	}
+
+	/* Only when post_write callback is set, we allow larger content than our
+	 * buffer sizes. The post-write callback handles assembling of the data
+	 */
+	if (!res->post_write_cb) {
+		if (offset > data_len) {
```

Comment:
Actually, this suggestion was what I had here earlier, it caused interopt tests to fail.

On some test cases, we end up writing an empty string resource with an empty string. And it should work.
Therefore sometimes there are cases where offset == 0 and data_len == 0. These are handled in the content format parser itself.

Reply:
Just a couple of suggestions:
- To allow the test cases where  offset == 0 and data_len == 0:
```suggestion
		if (offset >= data_len && data_len > 0) {
```
- Considering that the offset can be non-zero only in case we are doing a block-wise transfer (thus total_size is set), rather than accepting the first blocks that can fit in `data_len`, shouldn't we abort?
```suggestion
		if ((offset >= data_len && data_len > 0) || (total_size > data_len)) {
```

Reply:
Until there are proper tests for this, I would not like to change the logic.
For all size related checks, I would like to push those into the content-formats.

Reply:
Actually.. After a second though, I changed this to
`if ((offset > 0 && offset >= data_len) || total_size > data_len) {`